src/enumerable/enumerable_gather.jl

```diff
@@ -0,0 +1,75 @@
+struct EnumerableGather{T,S,F,I} <: Enumerable
+    source::S
+    fields::F
+    indexFields::I
+    key::Symbol
+    value::Symbol
+end
+
+struct Not{T}
+    val::T
+end
+
+function gather(source::Enumerable, args...; key::Symbol = :key, value::Symbol = :value)
+    T = eltype(source)
```

Comment:
Can we make ``T`` hold the type of the elements that are generated from this iterator? Probably something like ``NamedTuple{(iter.key, iter.value, savedFields...)}``. If we need access to the element type of the source also later, we can add another type parameter to ``EnumerableGather`` that is named ``TSOURCE``.

Reply:
I found that the code for it resembles the one used in `eltype` for `EnumerableGather` a lot. Do we need to create a helper function and make both `eltype` and `gather` call this function?

Reply:
I think the code could just be here, then you store the element type that is returned as ``T``, and ``eltype`` simply becomes ``Base.eltype(iter::EnumerableGather{T, S, F, I}) where {T, S, F, I} = T``, right?